# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: security/manager/ssl/nsICertStorage.idl
# Commit: 16350e08d34f
# Full Hash: 16350e08d34ff06c7f46745c6b160c7a7bda9140
# Author: Dana Keeler <dkeeler@mozilla.com>
# Date: 2019-12-06 04:43:57
# Regressor Bug: 1586855
# File Overlap Count: 6
# Description:
#   bug 1586855 - incorporate CRLite filters into cert_storage r=jcj,kjacobs
#   
#   This patch implements CRLite lookups for TLS server certificate revocation
#   information in telemetry-only mode. It adds a new preference
#   "security.pki.crlite_mode" to control the behavior of this feature. Setting
# ==============================================================================

diff -r 64fd40663930 -r 16350e08d34f security/manager/ssl/nsICertStorage.idl
--- a/security/manager/ssl/nsICertStorage.idl	Thu Dec 05 22:23:54 2019 +0000
+++ b/security/manager/ssl/nsICertStorage.idl	Thu Dec 05 22:41:53 2019 +0000
@@ -91,6 +91,8 @@
   const octet DATA_TYPE_REVOCATION = 1;
   const octet DATA_TYPE_CERTIFICATE = 2;
   const octet DATA_TYPE_CRLITE = 3;
+  const octet DATA_TYPE_CRLITE_FILTER_FULL = 4;
+  const octet DATA_TYPE_CRLITE_FILTER_INCREMENTAL = 5;
 
   /**
    * Asynchronously check if the backing storage has stored data of the given
@@ -104,6 +106,7 @@
 
   const short STATE_UNSET = 0;
   const short STATE_ENFORCE = 1;
+  const short STATE_NOT_ENROLLED = 2;
 
   /**
    * Asynchronously set the revocation states of a set of certificates.
@@ -159,6 +162,35 @@
   short getCRLiteState(in Array<octet> subject, in Array<octet> spki);
 
   /**
+   * Given the contents of a CRLite filter and its creation date as seconds since the epoch,
+   * replaces any existing filter with the new one.
+   */
+  [must_use]
+  void setFullCRLiteFilter(in Array<octet> filter, in uint64_t timestamp,
+                           in nsICertStorageCallback callback);
+
+  /**
+   * Given the DER-encoded issuer distinguished name, DER-encoded issuer subject public key info,
+   * and the bytes of the value of the serial number (so, not including the DER tag and length) of a
+   * certificate, returns the result of looking up the corresponding entry in the currently-saved
+   * CRLite filter (if any). Returns STATE_ENFORCE if the lookup indicates the corresponding
+   * certificate is revoked via CRLite. Returns STATE_UNSET if the lookup indicates the
+   * corresponding certificate is not revoked via CRLite. Returns STATE_NOT_ENROLLED if the issuer
+   * certificate is not enrolled in CRLite and thus no lookup was made. Also returns the timestamp
+   * before which lookups will be valid. That is, if a certificate has a notBefore value after the
+   * returned filter timestamp, the lookup is not trustworthy because the certificate may have been
+   * created after the filter, and thus it may cause a false negative or positive. If no filter is
+   * available (it may not have been downloaded yet), validBefore will be 0. The timestamp is
+   * represented as seconds since the epoch (i.e. it returns what was most recently set via
+   * setFullCRLiteFilter).
+   */
+  [must_use]
+  short getCRLiteRevocationState(in Array<octet> issuer,
+                                 in Array<octet> issuerSPKI,
+                                 in Array<octet> serialNumber,
+                                 out uint64_t validBefore);
+
+  /**
    * Trust flags to use when adding a adding a certificate.
    * TRUST_INHERIT indicates a certificate inherits trust from another
    * certificate.